src/PrestaShopBundle/Form/Admin/Improve/International/Geolocation/GeolocationFormDataProvider.php

```diff
@@ -0,0 +1,138 @@
+
+    /**
+     * @var GeoLiteCityCheckerInterface
+     */
+    private $geoLiteCityChecker;
+
+    /**
+     * @var Validate
+     */
+    private $validate;
+
+    /**
+     * @param DataConfigurationInterface $geolocationByIpAddressConfiguration
+     * @param DataConfigurationInterface $geolocationIpAddressWhitelistConfiguration
+     * @param DataConfigurationInterface $geolocationOptionsConfiguration
+     * @param GeoLiteCityCheckerInterface $geoLiteCityChecker
+     * @param Validate $validate
+     */
+    public function __construct(
+        DataConfigurationInterface $geolocationByIpAddressConfiguration,
+        DataConfigurationInterface $geolocationIpAddressWhitelistConfiguration,
+        DataConfigurationInterface $geolocationOptionsConfiguration,
+        GeoLiteCityCheckerInterface $geoLiteCityChecker,
+        Validate $validate
+    ) {
+        $this->geolocationByIpAddressConfiguration = $geolocationByIpAddressConfiguration;
+        $this->geolocationIpAddressWhitelistConfiguration = $geolocationIpAddressWhitelistConfiguration;
+        $this->geolocationOptionsConfiguration = $geolocationOptionsConfiguration;
+        $this->geoLiteCityChecker = $geoLiteCityChecker;
+        $this->validate = $validate;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData()
+    {
+        return [
+            'geolocation_by_id_address' => $this->geolocationByIpAddressConfiguration->getConfiguration(),
+            'geolocation_ip_address_whitelist' => $this->geolocationIpAddressWhitelistConfiguration->getConfiguration(),
+            'geolocation_options' => $this->geolocationOptionsConfiguration->getConfiguration(),
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setData(array $data)
+    {
+        $errors = [];
+
+        if ($data['geolocation_by_id_address']['geolocation_enabled'] && !$this->geoLiteCityChecker->isAvailable()) {
+            $errors[] = [
+                'key' => 'The geolocation database is unavailable.',
+                'parameters' => [],
+                'domain' => 'Admin.International.Notification',
+            ];
+        }
+
+        if (empty($data['geolocation_options']['geolocation_countries'])) {
+            $errors[] = [
+                'key' => 'Country selection is invalid.',
+                'parameters' => [],
+                'domain' => 'Admin.International.Notification',
+            ];
+        }
+
+        if (!$this->validate->isCleanHtml($data['geolocation_ip_address_whitelist']['geolocation_whitelist'])) {
+            $errors[] = [
+                'key' => 'Invalid whitelist',
+                'parameters' => [],
+                'domain' => 'Admin.International.Notification',
+            ];
+        }
+
+        if (!empty($errors)) {
+            return $errors;
+        }
+
+        return array_merge(
+            $this->geolocationByIpAddressConfiguration->updateConfiguration($data['geolocation_by_id_address']),
```

Comment:
the tests you are doing here... should be done in geolocationByIpAddressConfiguration class, am I right?

so in setData here, you should only have 

```php
return array_merge(
    $this->geolocationByIpAddressConfiguration->updateConfiguration($data['geolocation_by_id_address']),
    ...
);
```

Reply:
Well, `setData()` handles all form data, so if at least one field is invalid I don't want to save only valid fields but show error instead. If I were to delegate saving to each configuration service, there can be situation when 2 data configurators saves data and 1 fails due to validation. 

Another option would be to use something like:
```php
$errors = $this->geolocationByIpAddressConfiguration->validateConfiguration($data['geolocation_by_id_address']);
$errors += $this->geolocationOptionsConfiguration->validateConfiguration($data['geolocation_options']);

if (empty($errors)) {
    return array_merge(
          $this->geolocationByIpAddressConfiguration->updateConfiguration($data['geolocation_by_id_address']),
        ...
    );
}

return $errors;
```

but the method `validateConfiguration()` of contract `DataConfigurationInterface` does not really allow that:
```php
    /**
     * @return bool Returns true if no exception are thrown
     */
    public function validateConfiguration(array $configuration);
```

We can get only `true` or `false` after validation, but not error messages. Could we update contract to this?
```diff
+ @return array Errors if any or empty array
- @return bool Returns true if no exception are thrown
```

What do you think?